### PR TITLE
CreateExpressMiddleware: Send 404 code for disabled feature

### PR DESCRIPTION
**Pull request by jwicks31**
I'm not sure if this is the correct way to handle it, but I think we need to send the 404 status after setting it. Otherwise the status can be overwritten in the handler.

### CODE REVIEW

File: src/create-express-middleware.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { mergeFeatureNames } from './merge-feature-names';
 import { getQueryFeatureNames } from './get-query-feature-names';
 
 const setStatus = (res, isActiveFeatureName) =>
-  isActiveFeatureName ? res.status(200) : res.status(404);
+  isActiveFeatureName ? res.status(200) : res.status(404).send();
```
Comment (jwicks31) [inline src/create-express-middleware.js]: Should we set the status here at all if the feature is enabled?

Comment (ericelliott) [reply src/create-express-middleware.js]: If this is the only place the status gets set...

Comment (ericelliott) [reply src/create-express-middleware.js]: What do you think?

Comment (ericelliott) [reply src/create-express-middleware.js]: Maybe the line that calls this should be:

```
if (!isActiveFeatureName(requiredFeature, updatedFeatures) {
  return res.status(404).send();
}
```

Don't call the handler. Don't call `next()`.

**Comment by ericelliott** [comment]
LGTM, feel free to merge.

**Comment by ericelliott** [comment]
Express lets you set up error handlers that are automatically sent when you set the status and call `.send()`

**Comment by ericelliott** [comment]
Looks like you're right. We need to take the 404 page as a parameter, and send it with the `.send()` call.

**Comment by kennylavender** [comment]
Instead of taking a 404 page we could take a disabled handler to be more flexible. I don't think that will work with nextjs though, I think the only way to to get nextjs to render a 404 is to set the status from getInitialProps like you pointed out in the other repo. So maybe we can create some sort of nextjs specific HOC for feature toggles.

**Comment by jwicks31** [comment]
@kennylavender could you help me with what you mean by accepting a disabled handler? 

Will this need to behave differently for routes that render pages than for routes that are meant to send JSON responses? 

Something like:

```js
if (!isActiveFeatureName(requiredFeature, updatedFeatures)) {
      res.status(404);
      if (req.accepts('html')) return res.render(404Page,  { url: req.url })
      else return res.send({ error: 'Not found' });
    }
```

**Comment by kennylavender** [comment]
@jwicks31 our middleware shouldn't know about accept types, that responsibility should belong to the provided handlers.

One way we could allow the user to provide a disabled handler is to have them provide an enabled and disabled handler for each method.

```
const features = []
app.use(createExpressMiddleware({ features }, 'help', {
  get: {
    enabled: () => {},
    disabled: () => {},
  }, 
  post: {
    enabled: () => {},
    disabled: () => {},
  }
}))
```

I don't really care for this idea though, as a user I would not want to do this. In my opinion this middleware needs to be simplified/adjusted, if I remember correctly that is why I did not document it the middleware, I didn't want people using it as is. I am thinking this middleware should be adjusted to take a single enabled and disabled handler.

I wrote a feature toggle middleware for one of the po.et projects, I can't remember which, that I think was better. I will see if I can find it or write out a new example.

**Comment by kennylavender** [comment]
Can't find the middleware in the po.et projects.

But I think I would do something like this

```js
createExpressMiddleware(disabledHandler: Function) => (enbaledHandler: Function) => ({ requiredFeatures: [ ...String], features: [ ...Feature ]})
```

The user can then use the more specific `app.get`, `app.post` etc. and they can still use `app.use` when their handlers do the method switching themselfs.

```js
import { createExpressMiddleware } from '@parralleldrive/feature-toggles'

const notFoundHandler = (res, req, next) => {}

const todosHandler = (res, req, next) => {}

const features = []

app.get(
  '/todos',
  createExpressMiddleware(
    notFoundHandler,
    todosHandler,
    { requiredFeatures: ['todos'], features }
  )
)
```

**Comment by ericelliott** [comment]
That looks pretty good. If we auto-curry, it would be trivial to configure a featureOr404 handler.

Why is `requiredFeatures` an array in this example? If features can have dependencies, couldn't you create a single feature name that encapsulates all of the dependent features?

**Comment by kennylavender** [comment]
> That looks pretty good. If we auto-curry, it would be trivial to configure a featureOr404 handler.

Yup!

> Why is `requiredFeatures` an array in this example? If features can have dependencies, couldn't you create a single feature name that encapsulates all of the dependent features?

Yup, that works too.

**Comment by ericelliott** [comment]
> Yup, that works too. I think I was favoring avoiding dependencies and keeping features flat.

The alternative is that you then have to keep track of the feature graph everywhere you use features throughout the app, which multiplies the complexity.

**Pull request closed without merging**